File: design_patterns/002_observer/solutions/python/weatherStation/src/DisplayStatistic.py
from IDisplay import IDisplay
from IObserver import IObserver
from IPublisher import IPublisher
from WeatherData import WeatherInfo


class DisplayStatistic(IObserver, IDisplay):
    _weather_data:IPublisher
    _max_temp:float = 0.0
    _min_temp:float = 200.0
    _temp_sum:float = 0.0
    _num_readings:int = 0


    def __init__(self, weather_data:IPublisher) -> None:
        self._weather_data = weather_data
        self._weather_data.subscribe(self)


    def update(self, updated_data: dict) -> None:
        if current_temp := updated_data.get(WeatherInfo.TEMPERATURE):
            self._temp_sum += current_temp
            self._num_readings += 1
            if current_temp > self._max_temp:
                self._max_temp = current_temp
            if current_temp < self._min_temp:
                self._min_temp = current_temp
            self.display()


    def display(self) -> None:
        print(f'Avg/Max/Min temperature = {self._temp_sum/self._num_readings}/{self._max_temp}/{self._min_temp}')


if __name__ == '__main__':
    raise Exception(f'This module should not be executed directly. Only for imports')

File: design_patterns/002_observer/solutions/typescript/weatherStation/src/DisplayStatistic.ts
import { IDisplay } from './IDisplay';
import { IObserver } from './IObserver';
import { IPublisher } from './IPublisher';
import { WeatherInfo } from './WeatherData';

export class DisplayStatistic implements IObserver, IDisplay {
  private _weather_data: IPublisher;
  private _max_temp: number = 0.0;
  private _min_temp: number = 200.0;
  private _temp_sum: number = 0.0;
  private _num_readings: number = 0;

  constructor(weather_data: IPublisher) {
    this._weather_data = weather_data;
    weather_data.subscribe(this);
  }

  update(updated_data: { [key: string]: any }): void {
    const current_temp = updated_data[WeatherInfo.TEMPERATURE];

    if (current_temp !== undefined) {
      this._temp_sum += current_temp;
      this._num_readings++;

      if (current_temp > this._max_temp) {
        this._max_temp = current_temp;
      }

      if (current_temp < this._min_temp) {
        this._min_temp = current_temp;
      }

      this.display();
    }
  }

  display(): void {
    const avgTemp = this._temp_sum / this._num_readings;
    console.log(`Avg/Max/Min temperature = ${avgTemp}/${this._max_temp}/${this._min_temp}`);
  }
}
